upgrade/upgrade-scripts/upgrade

```diff
@@ -395,16 +397,21 @@ function finalize() {
 		;;
 	esac
 
-	ROOTFS_DATASET=$(get_mounted_rootfs_container_dataset)
-	[[ -n "$ROOTFS_DATASET" ]] ||
-		die "unable to determine mounted rootfs container dataset"
-
-	SNAPSHOT_NAME=$(get_dataset_rollback_snapshot_name "$ROOTFS_DATASET")
-	[[ -n "$SNAPSHOT_NAME" ]] ||
-		die "unable to determine rollback snapshot name"
-
-	zfs destroy -r "$ROOTFS_DATASET@$SNAPSHOT_NAME" ||
-		die "failed to destroy rollback snapshots"
+	SNAPSHOT_LIST=$(zfs list -t snapshot -r rpool/ROOT -d 2 -Ho name) ||
+		die "unable to determine available rollback snapshots"
+
+	for snapshot in $SNAPSHOT_LIST; do
```

Comment:
Since we're assuming each snapshot in the list here, is one of the `execute-upgrade.XXXXXXX` snapshots.. would it be reasonable to add some verification of the snapshot name in the block below? e.g. something like this:
```
grep -qE "^execute-upgrade.[[:alnum:]]{7}$" <<< "$snapshot" || die "unexpected snapshot name: '$snapshot'"
```
This might help us catch the case where/if we start adding other snapshots to the container dataset(s), but forget to update this cleanup code.. in that case, this check would fail, so we'd (hopefully) more quickly notice the problem.

Reply:
That's a good idea, I can include that

Reply:
May I suggest we warn instead of die here. Feels like an unexpected snapshot should not prevent us from completing the operation. Another concern is if we do die here, the stackStartup will hit a DFE.

Reply:
> May I suggest we warn instead of die here.

I disagree. IMO, warnings are generally useless, since they mostly go ignored. So, if we went that route, I'd opt to do nothing, rather than warn (we will already log the removal command in syslog).

With that said, the point of this check is to avoid removing a snapshots that we didn't not expect to remove; e.g. avoid "data corruption" where we accidentally remove a snapshot we should not have removed. So, in that case, a warning will not prevent us from removing that unexpected snapshot.

Additionally, we should never hit this case. If we do, it's a bug, and I'd rather halt the operation immediately, rather than assume it's OK to remove an arbitrary snapshot we weren't expecting and carry on.

> Another concern is if we do die here, the stackStartup will hit a DFE.

This is my intention.

Reply:
Hmm my intention was we would warn and skip that snapshot but I see your point about it being a bug from a product standpoint. Are you aware if support engineers ever take snapshots? say when installing hotfixes or doing some manual intervention on the engine.

Reply:
> Hmm my intention was we would warn and skip that snapshot

I'd be more OK with that approach, but I still think it's a little strange to skip the snapshot without having some prior knowledge about what the snapshot would contain, and why it'd be better to skip it rather than fail.

> Are you aware if support engineers ever take snapshots?

I am not, but I also haven't tried to ask around to find out.. this might be a good question for Anthony.

Reply:
I just asked Anthony for some input, will update you on the response

Reply:
Anthony's response
```
Hey Aditya, not normally. It's possible we could if we're looking into a specific issue just to preserve the state but we try not to do anything the stack wouldn't be aware of.
So it would probably be fairly short-lived if we did it.
```

Reply:
Good to know. Thanks. Changes look good.